cpp/src/bitmask/bitmask_util.cuh

```diff
@@ -0,0 +1,324 @@
+/*
+ * Copyright (c) 2018, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+	 * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef _BITMASK_UTIL_H_
+#define _BITMASK_UTIL_H_
+
+#pragma once
+
+
+
+#include "cudf.h"
+#include "rmm/rmm.h"
+#include "utilities/error_utils.h"
+#include <cuda_runtime_api.h>
+
+
+namespace bitmask {
+
+//----------------------------------------------------------------------------------------------------------
+//			Host and Device Utility Functions
+//----------------------------------------------------------------------------------------------------------
+namespace util {
+
+	/**
+	 * determine the bitmap that contains a record
+	 * @param[in]  record_idx    The record index
+	 * @return the bitmap index
+	 */
+	__host__ __device__  int which_bitmap_record(int record_idx) {
+		return (record_idx / GDF_VALID_BITSIZE);
+	}
+
+	/**
+	 * determine which bit in a bitmap relates to a record
+	 * @param[in]  record_idx    The record index
+	 * @return which bit within the bitmap
+	 */
+	__host__ __device__ int which_bit(int record_idx) {
+		return (record_idx % GDF_VALID_BITSIZE);
+	}
+
+}	// end util namespace
+
+
+
+//----------------------------------------------------------------------------------------------------------
+//			Device Only Utility Functions
+//----------------------------------------------------------------------------------------------------------
+namespace device {
+
+	/**
+	 * Check to see if a record is Valid (aka not null)
+	 *
+	 * @param[in] valid        the device memory containing the valid bitmaps
+	 * @param[in] record_idx   the record index to check
+	 *
+	 * @return  true if record is valid, false if record is null
+	 */
+	__device__ bool is_valid(gdf_valid_type * valid, int record_idx) {
+		int rec = bitmask::util::which_bitmap_record(record_idx);
+		int bit = bitmask::util::which_bit(record_idx);
+
+		int status = atomicAnd(&valid[rec], (1U << bit));
+
+		return ( status == 0) ? false : true;
+
+	}
+
+
+	/**
+	 * Set a bit
+	 *
+	 * @param[in] valid         the valid memory array
+	 * @param[in] record_idx    the record index
+	 *
+	 * @return gdf_error - did it work or not
+	 */
+	__device__ gdf_error set_bit(gdf_valid_type * valid, int record_idx) {
+		int rec = bitmask::util::which_bitmap_record(record_idx);
+		int bit = bitmask::util::which_bit(record_idx);
+
+		atomicOr( &valid[rec],   (1U << bit));		// set the bit
```

Comment:
Just a bit safer. In case `gdf_valid_type` ever becomes a 8B type.
```suggestion
		atomicOr( &valid[rec],   (gdf_valid_type{1} << bit));		// set the bit
```

Reply:
good catch..
The `1U`  does introducing the same issues that we are trying to over come, which is to allow the underlying data type to change